src/python/zquantum/core/wip/cost_function.py

```diff
@@ -0,0 +1,260 @@
+            protocol and are used to create the estimation tasks.
+        fixed_parameters: values for the circuit parameters that should be fixed.
+        parameter_precision: the standard deviation of the Gaussian noise to add to each parameter, if any.
+        parameter_precision_seed: seed for randomly generating parameter deviation if using parameter_precision
+        gradient_function: a function which returns a function used to compute the gradient of the cost function
+            (see from zquantum.core.gradients.finite_differences_gradient for reference)
+
+    Returns:
+        Callable
+    """
+    estimation_tasks = [
+        EstimationTask(
+            operator=target_operator, circuit=parametrized_circuit, number_of_shots=None
+        )
+    ]
+
+    if estimation_preprocessors is None:
+        estimation_preprocessors = []
+
+    for estimation_preprocessor in estimation_preprocessors:
+        estimation_tasks = estimation_preprocessor(estimation_tasks)
+
+    circuit_symbols = sorted(
+        list(
+            set(
+                [
+                    param
+                    for task in estimation_tasks
+                    for param in task.circuit.symbolic_params
+                ]
+            )
+        ),
+        key=str,
+    )
+
+    def ground_state_cost_function(
+        parameters: np.ndarray, store_artifact: StoreArtifact = None
+    ) -> ValueEstimate:
+        """Evaluates the expectation value of the op
+
+        Args:
+            parameters: parameters for the parameterized quantum circuit
+
+        Returns:
+            value: estimated energy of the target operator with respect to the circuit
+        """
+        nonlocal estimation_tasks
+        parameters = parameters.copy()
+        if fixed_parameters is not None:
+            parameters = combine_ansatz_params(fixed_parameters, parameters)
+        if parameter_precision is not None:
+            rng = np.random.default_rng(parameter_precision_seed)
+            noise_array = rng.normal(0.0, parameter_precision, len(parameters))
+            parameters += noise_array
+
+        symbols_map = create_symbols_map(circuit_symbols, parameters)
+        estimation_tasks = evaluate_estimation_circuits(
+            estimation_tasks, [symbols_map for _ in estimation_tasks]
+        )
+
+        expectation_values_list = estimator(backend, estimation_tasks)
+        summed_values = ValueEstimate(
+            np.sum(
+                [
+                    np.sum(expectation_values.values)
+                    for expectation_values in expectation_values_list
+                ]
+            )
+        )
+
+        return summed_values
+
+    return function_with_gradient(
+        ground_state_cost_function, gradient_function(ground_state_cost_function)
+    )
+
+
+def sum_expectation_values(expectation_values: ExpectationValues) -> ValueEstimate:
```

Comment:
Agreed. It should probably be moved to `zquantum.core.utils`, but I would prefer to wait until the wip is merged into core since it's a relatively low-impact switch. Putting a chore into tracker

Reply:
https://www.pivotaltracker.com/story/show/177854476

Reply:
Sounds fine!